File: sdks/local-rp/ocaml/lib/timeutil.ml
(* RFC3339 timestamp parsing shared by every module that checks freshness.

   Every "current time" in this SDK is an explicit [float] (Unix epoch
   seconds, UTC) parameter, never read from the system clock internally --
   mirroring [liblinkkeys]'s discipline of taking [now] as an argument so
   verification stays deterministic and testable (the sole exception is
   [Revocation.count_valid_signers]'s wall-clock default, documented there,
   matching the Rust reference's [check_signing_key_valid]).

   OCaml's stdlib has no RFC3339 parser and [Unix.mktime] is local-time-zone
   dependent (unsuitable for a portable, deterministic UTC parse), so this
   module hand-rolls both directions using Howard Hinnant's [days_from_civil]
   / [civil_from_days] algorithm -- a well-known, TZ-independent, proleptic
   Gregorian <-> day-count conversion that is exact under C-style truncating
   integer division (which is also OCaml's [/]). *)

exception Bad_timestamp of string

(* Days since 1970-01-01 for the given proleptic-Gregorian civil date. *)
let days_from_civil (y : int) (m : int) (d : int) : int =
  let y = if m <= 2 then y - 1 else y in
  let era = (if y >= 0 then y else y - 399) / 400 in
  let yoe = y - (era * 400) in
  (* [0, 399] *)
  let mp = (m + 9) mod 12 in
  (* [0, 11] *)
  let doy = ((153 * mp) + 2) / 5 + d - 1 in
  (* [0, 365] *)
  let doe = (yoe * 365) + (yoe / 4) - (yoe / 100) + doy in
  (* [0, 146096] *)
  (era * 146097) + doe - 719468

let civil_from_days (z : int) : int * int * int =
  let z = z + 719468 in
  let era = (if z >= 0 then z else z - 146096) / 146097 in
  let doe = z - (era * 146097) in
  (* [0, 146096] *)
  let yoe = (doe - (doe / 1460) + (doe / 36524) - (doe / 146096)) / 365 in
  (* [0, 399] *)
  let y = yoe + (era * 400) in
  let doy = doe - ((365 * yoe) + (yoe / 4) - (yoe / 100)) in
  (* [0, 365] *)
  let mp = ((5 * doy) + 2) / 153 in
  (* [0, 11] *)
  let d = doy - (((153 * mp) + 2) / 5) + 1 in
  (* [1, 31] *)
  let m = mp + if mp < 10 then 3 else -9 in
  (* [1, 12] *)
  let y = if m <= 2 then y + 1 else y in
  (y, m, d)

let fail fmt = Printf.ksprintf (fun s -> raise (Bad_timestamp s)) fmt

(* Parse an RFC3339 timestamp into Unix epoch seconds (UTC, as a float so
   fractional seconds survive). Requires an explicit timezone offset (Z or
   +HH:MM/-HH:MM) -- a bare local-time string is rejected rather than
   silently treated as UTC or local time. *)
let parse_rfc3339 (raw : string) : float =
  let s = String.trim raw in
  let len = String.length s in
  if len < 20 then fail "timestamp too short: %S" s;
  (try
     if s.[4] <> '-' || s.[7] <> '-' then fail "bad date separators: %S" s;
     if s.[10] <> 'T' && s.[10] <> 't' then fail "missing T separator: %S" s;
     let year = int_of_string (String.sub s 0 4) in
     let month = int_of_string (String.sub s 5 2) in
     let day = int_of_string (String.sub s 8 2) in
     let rest = String.sub s 11 (len - 11) in
     if String.length rest < 9 then fail "time part too short: %S" s;
     if rest.[2] <> ':' || rest.[5] <> ':' then fail "bad time separators: %S" s;
     let hour = int_of_string (String.sub rest 0 2) in
     let minute = int_of_string (String.sub rest 3 2) in
     let sec = int_of_string (String.sub rest 6 2) in
     let after_sec = String.sub rest 8 (String.length rest - 8) in
     let frac, after_frac =
       if String.length after_sec > 0 && after_sec.[0] = '.' then begin
         let idx = ref 1 in
         while
           !idx < String.length after_sec
           && (let c = after_sec.[!idx] in
               c >= '0' && c <= '9')
         do
           incr idx
         done;
         let digits = String.sub after_sec 1 (!idx - 1) in
         let frac = if digits = "" then 0.0 else float_of_string ("0." ^ digits) in
         (frac, String.sub after_sec !idx (String.length after_sec - !idx))
       end
       else (0.0, after_sec)
     in
     if after_frac = "" then fail "timestamp has no timezone offset: %S" s;
     let offset_seconds =
       if after_frac = "Z" || after_frac = "z" then 0
       else begin
         let sign =
           match after_frac.[0] with
           | '+' -> 1
           | '-' -> -1
           | _ -> fail "timestamp has no timezone offset: %S" s
         in
         let body = String.sub after_frac 1 (String.length after_frac - 1) in
         let body =
           if String.length body = 4 && not (String.contains body ':') then
             String.sub body 0 2 ^ ":" ^ String.sub body 2 2
           else body
         in
         if String.length body <> 5 || body.[2] <> ':' then fail "bad offset: %S" s;
         let oh = int_of_string (String.sub body 0 2) in
         let om = int_of_string (String.sub body 3 2) in
         sign * ((oh * 3600) + (om * 60))
       end
     in
     if month < 1 || month > 12 || day < 1 || day > 31 || hour > 23 || minute > 59 || sec > 60 then
       fail "field out of range: %S" s;
     let days = days_from_civil year month day in
     (float_of_int ((days * 86400) + (hour * 3600) + (minute * 60) + sec - offset_seconds)) +. frac
   with
  | Bad_timestamp _ as e -> raise e
  | Failure msg -> fail "unparseable RFC3339 timestamp %S: %s" s msg
  | Invalid_argument msg -> fail "unparseable RFC3339 timestamp %S: %s" s msg)

(* Render Unix epoch seconds as RFC3339 UTC with a Z suffix. Not
   wire-normative (only the parsed instant is ever compared -- see the
   design doc's Wire Precision), but whole-second precision is used
   whenever there is no meaningful sub-second component. *)
let to_rfc3339 (t : float) : string =
  let days = int_of_float (Float.floor (t /. 86400.)) in
  let day_start = float_of_int days *. 86400. in
  let rem = t -. day_start in
  let y, m, d = civil_from_days days in
  let total_sec = int_of_float (Float.floor rem) in
  let frac = rem -. float_of_int total_sec in
  let hh = total_sec / 3600 in
  let mm = total_sec mod 3600 / 60 in
  let ss = total_sec mod 60 in
  if frac < 1e-9 then Printf.sprintf "%04d-%02d-%02dT%02d:%02d:%02dZ" y m d hh mm ss
  else begin
    let micros = int_of_float (Float.round (frac *. 1e6)) in
    if micros >= 1_000_000 then Printf.sprintf "%04d-%02d-%02dT%02d:%02d:%02dZ" y m d hh mm (ss + 1)
    else Printf.sprintf "%04d-%02d-%02dT%02d:%02d:%02d.%06dZ" y m d hh mm ss micros
  end

File: sdks/local-rp/c/src/time_util.c
#include "time_util.h"

#include <stdio.h>
#include <string.h>

#include "error.h"

int64_t lrp_days_from_civil(int64_t y, unsigned m, unsigned d) {
    y -= (m <= 2);
    int64_t era = (y >= 0 ? y : y - 399) / 400;
    unsigned yoe = (unsigned)(y - era * 400);              /* [0, 399] */
    unsigned doy = (153 * (m + (m > 2 ? (unsigned)-3 : 9)) + 2) / 5 + d - 1; /* [0, 365] */
    unsigned doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;   /* [0, 146096] */
    return era * 146097 + (int64_t)doe - 719468;
}

static void civil_from_days(int64_t z, int64_t *y, unsigned *m, unsigned *d) {
    z += 719468;
    int64_t era = (z >= 0 ? z : z - 146096) / 146097;
    unsigned doe = (unsigned)(z - era * 146097);            /* [0, 146096] */
    unsigned yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365; /* [0, 399] */
    int64_t yy = (int64_t)yoe + era * 400;
    unsigned doy = doe - (365 * yoe + yoe / 4 - yoe / 100); /* [0, 365] */
    unsigned mp = (5 * doy + 2) / 153;                      /* [0, 11] */
    unsigned dd = doy - (153 * mp + 2) / 5 + 1;              /* [1, 31] */
    unsigned mm = mp + (mp < 10 ? 3 : (unsigned)-9);
    yy += (mm <= 2);
    *y = yy;
    *m = mm;
    *d = dd;
}

static int parse_uint_fixed(const char *s, size_t n, unsigned *out) {
    unsigned v = 0;
    for (size_t i = 0; i < n; i++) {
        if (s[i] < '0' || s[i] > '9') return -1;
        v = v * 10 + (unsigned)(s[i] - '0');
    }
    *out = v;
    return 0;
}

int lrp_parse_rfc3339(const char *s, int64_t *out_unix, lrp_error *err) {
    size_t len = strlen(s);
    /* Minimum: YYYY-MM-DDTHH:MM:SSZ = 20 chars */
    if (len < 20 || s[4] != '-' || s[7] != '-' || (s[10] != 'T' && s[10] != 't' && s[10] != ' ') ||
        s[13] != ':' || s[16] != ':') {
        return lrp_fail(err, LRP_ERR_DECODE, "invalid timestamp: bad format");
    }
    unsigned year4, month, day, hour, minute, second;
    if (parse_uint_fixed(s, 4, &year4) != 0 || parse_uint_fixed(s + 5, 2, &month) != 0 ||
        parse_uint_fixed(s + 8, 2, &day) != 0 || parse_uint_fixed(s + 11, 2, &hour) != 0 ||
        parse_uint_fixed(s + 14, 2, &minute) != 0 || parse_uint_fixed(s + 17, 2, &second) != 0) {
        return lrp_fail(err, LRP_ERR_DECODE, "invalid timestamp: non-numeric field");
    }
    if (month < 1 || month > 12 || day < 1 || day > 31 || hour > 23 || minute > 59 || second > 60) {
        return lrp_fail(err, LRP_ERR_DECODE, "invalid timestamp: field out of range");
    }

    size_t i = 19;
    /* Optional fractional seconds: .digits (truncated, not used). */
    if (i < len && s[i] == '.') {
        i++;
        while (i < len && s[i] >= '0' && s[i] <= '9') i++;
    }

    int64_t offset_seconds = 0;
    if (i >= len) {
        return lrp_fail(err, LRP_ERR_DECODE, "invalid timestamp: missing timezone");
    }
    if (s[i] == 'Z' || s[i] == 'z') {
        i++;
    } else if (s[i] == '+' || s[i] == '-') {
        int sign = (s[i] == '-') ? -1 : 1;
        i++;
        if (len - i < 5 || s[i + 2] != ':') {
            return lrp_fail(err, LRP_ERR_DECODE, "invalid timestamp: bad timezone offset");
        }
        unsigned oh, om;
        if (parse_uint_fixed(s + i, 2, &oh) != 0 || parse_uint_fixed(s + i + 3, 2, &om) != 0) {
            return lrp_fail(err, LRP_ERR_DECODE, "invalid timestamp: bad timezone offset");
        }
        offset_seconds = sign * (int64_t)(oh * 3600 + om * 60);
        i += 5;
    } else {
        return lrp_fail(err, LRP_ERR_DECODE, "invalid timestamp: bad timezone marker");
    }
    if (i != len) {
        return lrp_fail(err, LRP_ERR_DECODE, "invalid timestamp: trailing characters");
    }

    int64_t days = lrp_days_from_civil((int64_t)year4, month, day);
    int64_t secs = days * 86400 + (int64_t)hour * 3600 + (int64_t)minute * 60 + (int64_t)second;
    secs -= offset_seconds;
    *out_unix = secs;
    return 0;
}

void lrp_format_rfc3339(int64_t unix_seconds, char out[32]) {
    int64_t days = unix_seconds >= 0 ? unix_seconds / 86400 : -((-unix_seconds + 86399) / 86400);
    int64_t rem = unix_seconds - days * 86400;
    if (rem < 0) {
        rem += 86400;
        days -= 1;
    }
    unsigned hour = (unsigned)(rem / 3600);
    unsigned minute = (unsigned)((rem % 3600) / 60);
    unsigned second = (unsigned)(rem % 60);
    int64_t y;
    unsigned m, d;
    civil_from_days(days, &y, &m, &d);
    snprintf(out, 32, "%04lld-%02u-%02uT%02u:%02u:%02u+00:00", (long long)y, m, d, hour, minute,
             second);
}
